### ISSUE TITLE
EF Core 1.1 generating unique index on FK one-to-one

### ISSUE THREAD

**Issue by moraleslos** [dotnet/efcore#7249]
I'm using EF Core 1.1 and created my one-to-one relationships, an example:

```
EntityTypeBuilder<Account>
   .HasOne(a => a.Customer)
   .WithOne(a => a.Account)
   .HasForeignKey<Account>(a => a.CustomerId)
```

Now using Add-Migration on my DbContext it generates the following:

```
migrationBuilder.CreateIndex(
   name: "IX_Accounts_CustomerId",
   table: "Accounts",
   column: "CustomerId",
   unique: true);
```

Is there a way to specify in Fluent to remove the uniqueness from the index?  I don't mind the index being generated, but for my purposes it can not be unique.  The reason is that these tables are auditable so there can be duplicate data (outside of the PK).  Hence records can contain the same FK id if there were no changes to that FK entity (but maybe changes elsewhere in the record).  

TL;DR Any way of making the index non-unique in Fluent for one-to-one FKs?

**Comment by smitpatel** [comment]
Answered on stackoverflow. http://stackoverflow.com/questions/41170325/ef-core-1-1-generating-unique-index-on-fk-one-to-one/41172323#41172323

**Comment by moraleslos** [comment]
The SO answer didn't work for me.  It still created the unique index.

**Comment by moraleslos** [comment]
Can we add a feature for one-to-one mapping where we can designate the FK to be non-unique-index and if this is the case, EF would automatically restrict cascade deletes?

**Comment by moraleslos** [comment]
BTW, there is no OnModelConfiguring() so I assume you're talking about OnModelCreating(). If not, the only other method I see in DbContext is OnConfiguring(DbContextOptionsBuilder optionsBuilder), and I don't think this is what you're talking about.

**Comment by moraleslos** [comment]
I'll try-- it's a big app so I'll need to create a sample for you...

**Comment by smitpatel** [comment]
I have uploaded my sample app here https://github.com/smitpatel/EFSampleApp/tree/issue7249

**Comment by moraleslos** [comment]
I've uploaded a sample app that fails here:  [https://github.com/moraleslos/EFissue7249](url)

**Comment by smitpatel** [comment]
Filed #7254 - Change of PK on `Customer` is causing the configuration to lose. As a work-around you can use the fluent api to configure the index after you have configured `Customer` entity.

**Comment by moraleslos** [comment]
Thanks!  I'll follow up on issue 7254 then.  In the meantime I'll take your advice as a workaround.

**Comment by ajcvickers** [comment]
@moraleslos Just want to follow up on one thing: if EF pulls back more than one row with the same FK value, then the results will be undefined. EF may crash or it may corrupt data such that the wrong row may be updated. When a 1:1 relationship is defined, it is a requirement that no query ever return more than one entity on the dependent side of the relationship. Is the auditing you mention ensuring that this never happens?

**Comment by moraleslos** [comment]
@ajcvickers 

Yes the 1:1 is unique.  But because it's an audit table, there can me multiple records of the same exact data but with different audit ids, events (updated, added, deleted) and timestamps.  That's why I can't have that unique index applied to the FK.

Is there a way to define a 1:1 relationship in the DbContext without using HasOne().WithOne() so that I won't have to worry about the unique index?  

BTW I won't need opposite navigation properties.  Just need to be able to navigate from Account -> Customer. 

BTW2, @smitpatel's suggestion worked for me, although I have a feeling this is more of a "hack" just to get it to work.

**Comment by ajcvickers** [comment]
@moraleslos I think that this is an interesting definition of uniqueness that we hadn't considered--that is, at a conceptual level there aren't any duplicates and duplicates won't be returned from queries, but the column cannot be defined as unique because of the auditing. Does the same thing happen with PKs? Can they no longer have a unique constraint on them for the same reason?

I think that as a team we will need to discuss and probably investigate more.

Note that if you are using Migrations to create the database, then probably the best thing to do would be to edit the migration so that it doesn't create a unique index for the FK. This maintains the conceptual uniqueness in the model (which we are asserting is true) but relaxes it for table creation where the interaction with auditing makes it not true.

**Comment by moraleslos** [comment]
@ajcvickers 

I have mine set up "weird"... not sure if it's the right way of doing things.  I have the traditional data (EF) model with no auditing in one dbcontext and an extension (literally) of this model that includes auditing fields in another dbcontext (essentially add audit id, timestamp, username and event fields, where the audit id is now the PK).  Whenever the traditional EF model gets added/updated/deleted, I intercept that state and create a new auditing record with the same data (including PKs and FKs) but with event = state, timestamp = now, and autogenerated audit id.  

So the traditional data model is current while the auditing data model contains every thing [BTW, is ambient transactions ready yet?  =)]

I don't really perform any queries on the auditing database except for querying for a transaction history (via EF).  But either way the auditing tables mimic exactly that of the traditional tables, FKs and everything,  

Maybe my above design is incorrect and that is why EF can't really accommodate?

**Comment by ajcvickers** [comment]
@moraleslos We will discuss in the team--I don't know enough about auditing to know what best practices there are.

**Comment by moraleslos** [comment]
@ajcvickers There are some open source EF-based auditing frameworks out there that, in essence, apply the same concepts to what I just described.  For example: https://www.nuget.org/packages/TrackerEnabledDbContext 

However none that I've researched are EF Core compatible and I do need a little finer control of things hence they usually don't work out for my use cases.  But the concepts are very similar.  

Really EF Core doesn't need to do anything special for auditing.  I just need that 1:1 FK to be a non-unique index.  I tried a while ago to not define the 1:1 relationship for that audit dbcontext but it generated one anyway (I believe it's called shadowing)... so I just decided to explicitly define them myself.  That's why I suggested earlier if we could turn off the uniqueness-- or maybe even the index in general-- when defining the 1:1 relationship, e.g.

```
EntityTypeBuilder<Account>
   .HasOne(a => a.Customer)
   .WithOne(a => a.Account)
   .HasForeignKey<Account>(a => a.CustomerId)
   .HasIndex(false)
```

But it would be cool if you could build an add-on to EF Core that does this auditing OOTB... I'm sure lots of people would use it  ;-)

**Comment by divega** [comment]
**EF Team Triage:** We believe that the user creating a non-unique index explicitly is a reasonable way to assert that they don't want a unique index to be created. So we should do this. The auditing scenario could also be enabled with model-level filters.

**Comment by ajcvickers** [comment]
Closing this as #7254 was created to track this issue.

**Comment by ishakdemir** [comment]
> @ajcvickers There are some open source EF-based auditing frameworks out there that, in essence, apply the same concepts to what I just described. For example: https://www.nuget.org/packages/TrackerEnabledDbContext
> 
> However none that I've researched are EF Core compatible and I do need a little finer control of things hence they usually don't work out for my use cases. But the concepts are very similar.
> 
> Really EF Core doesn't need to do anything special for auditing. I just need that 1:1 FK to be a non-unique index. I tried a while ago to not define the 1:1 relationship for that audit dbcontext but it generated one anyway (I believe it's called shadowing)... so I just decided to explicitly define them myself. That's why I suggested earlier if we could turn off the uniqueness-- or maybe even the index in general-- when defining the 1:1 relationship, e.g.
> 
> ```
> EntityTypeBuilder<Account>
>    .HasOne(a => a.Customer)
>    .WithOne(a => a.Account)
>    .HasForeignKey<Account>(a => a.CustomerId)
>    .HasIndex(false)
> ```
> 
> But it would be cool if you could build an add-on to EF Core that does this auditing OOTB... I'm sure lots of people would use it ;-)

thank you
